File: Sort, Search, Tree1(BFT, DFT, BFS, DFS)/leetcode/quicksort/quicksort.py
#User function Template for python3

class Solution:
    #Function to sort a list using quick sort algorithm.
    def quickSort(self,arr,low,high):
        # code here
        if low < high:
            # partition 메서드를 호출하여 피벗의 위치를 얻음
            pivot_index = self.partition(arr, low, high)
            
            # 피벗을 기준으로 좌우 부분을 재귀적으로 정렬
            self.quickSort(arr, low, pivot_index - 1)
            self.quickSort(arr, pivot_index + 1, high)
            
    
    def partition(self,arr,low,high):
        # code here
        pivot = arr[low]  # 부분 리스트의 왼쪽 요소를 피벗으로 설정
        left = low + 1
        right = high

        # left가 right보다 작을 때만 반복
        while True:
            # left가 right보다 작고, arr[left]가 pivot보다 작거나 같은 동안 left 증가
            while left <= right and arr[left] <= pivot:
                left += 1
            
            # right가 left보다 크고, arr[right]가 pivot보다 큰 동안 right 감소
            while left <= right and arr[right] > pivot:
                right -= 1
            
            # 교환될 요소를 찾았다면 두 요소를 교환
            if left <= right:
                arr[left], arr[right] = arr[right], arr[left]
            else:
                break
        
        # 마지막으로 피벗 요소와 right가 가리키는 요소를 교환하여 피벗을 제자리로 이동
        arr[low], arr[right] = arr[right], arr[low]
        
        # 피벗이 위치한 인덱스를 반환
        return right
            
    


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == "__main__":
    t = int(input())
    for i in range(t):

        arr = list(map(int, input().split()))
        n = len(arr)
        Solution().quickSort(arr, 0, n - 1)
        for i in range(n):
            print(arr[i], end=" ")
        print()

# } Driver Code Ends

File: Sort, Search, Tree1(BFT, DFT, BFS, DFS)/leetcode/quicksort/quicksort.c
#include <stdio.h>

// 배열을 정렬하는 함수
void quickSort(int arr[], int low, int high);

// 배열을 분할하고 피벗의 위치를 반환하는 함수
int partition(int arr[], int low, int high);

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

// 퀵 정렬 함수
void quickSort(int arr[], int low, int high) {
    if (low < high) {
        // partition 함수 호출하여 피벗 위치 얻음
        int pivot_index = partition(arr, low, high);
        
        // 피벗을 기준으로 좌우 부분 배열을 재귀적으로 정렬
        quickSort(arr, low, pivot_index - 1);
        quickSort(arr, pivot_index + 1, high);
    }
}

// 분할 함수
int partition(int arr[], int low, int high) {
    int pivot = arr[low]; // 부분 리스트의 첫 번째 요소를 피벗으로 설정
    int left = low + 1;
    int right = high;

    while (1) {
        // left가 right보다 작고, arr[left]가 pivot보다 작거나 같은 동안 left 증가
        while (left <= right && arr[left] <= pivot) {
            left++;
        }

        // right가 left보다 크고, arr[right]가 pivot보다 큰 동안 right 감소
        while (left <= right && arr[right] > pivot) {
            right--;
        }

        // 교환될 요소를 찾았다면 두 요소를 교환
        if (left <= right) {
            swap(&arr[left], &arr[right]);
        } else {
            break;
        }
    }

    // 피벗을 제자리로 이동
    swap(&arr[low], &arr[right]);

    // 피벗의 최종 위치 반환
    return right;
}

// 메인 함수
int main() {
    int t;
    scanf("%d", &t); // 테스트 케이스 수 입력 받기

    for (int i = 0; i < t; i++) {
        int n;
        scanf("%d", &n); // 배열 크기 입력 받기

        int arr[n];
        for (int j = 0; j < n; j++) {
            scanf("%d", &arr[j]); // 배열 요소 입력 받기
        }

        quickSort(arr, 0, n - 1); // 퀵 정렬 수행

        for (int j = 0; j < n; j++) {
            printf("%d ", arr[j]); // 정렬된 배열 출력
        }
        printf("\n");
    }

    return 0;
}
